doc/source/data/training_ingest.rst

```diff
@@ -0,0 +1,525 @@
+            iterator_shards = datasets["train"].streaming_split(
+                world_size, equal=True, locality_hints=worker_node_ids
+            )
+
+            # Return the assigned iterators for each worker.
+            return [{"train": it} for it in iterator_shards]
+
+
+    my_trainer = TorchTrainer(
+        train_loop_per_worker,
+        scaling_config=ScalingConfig(num_workers=2),
+        datasets={"train": ds},
+        dataset_config=MyCustomDataConfig(),
+    )
+    my_trainer.fit()
+
+.. testoutput::
+    :hide:
+
+    ... 
+
+The subclass must be serializable, since it will be copied from the driver script to the driving actor of the Trainer. Its :meth:`configure <ray.train.DataConfig.configure>` method is called on the main actor of the Trainer group to create the data iterators for each worker.
+
+In general, you can use :class:`DataConfig <ray.train.DataConfig>` for any shared setup that has to occur ahead of time before the workers start iterating over data. The setup will be run at the start of each Trainer run.
+
+Performance tips
+~~~~~~~~~~~~~~~~
+
+.. _dataset_cache_performance:
+
+Caching the preprocessed Dataset
+################################
+If you are training on GPUs and have an expensive CPU preprocessing operation, this may bottleneck training throughput.
+
+If your preprocessed Dataset is small enough to fit in object store memory, the easiest thing to do is to *materialize* the preprocessed dataset in Ray object store memory, by calling :meth:`materialize() <ray.data.Dataset.materialize>` on the preprocessed dataset. This tells Ray Data to compute the entire preprocessed and pin it in the Ray object store memory. As a result, when iterating over the dataset repeatedly, the preprocessing operations do not need to be re-run. However, the trade-off is that if the preprocessed data is too large to fit into Ray object store memory, this will greatly decrease performance as data needs to be spilled to disk.
+
+.. testcode::
+
+    # Load the data.
+    train_ds = ray.data.read_parquet("s3://anonymous@ray-example-data/iris.parquet")
+
+    # Preprocess the data. Transformations that are made before the materialize call
+    # below will only be run once.
+    train_ds = train_ds.map_batches(normalize_length)
+
+    # Materialize the dataset in object store memory.
+    # Only do this if train_ds is small enough to fit in object store memory.
+    train_ds = train_ds.materialize()
+
+    # Add per-epoch preprocessing. Transformations that you want to run per-epoch, such
+    # as data augmentation, should go after the materialize call.
+    train_ds = train_ds.map_batches(augment_data)
+
+    # Pass train_ds to the Trainer as normal...
+
+Adding CPU-only nodes to your cluster
+#####################################
+If you are bottlenecked on expensive CPU preprocessing and the preprocessed Dataset is too large to fit in object store memory, then the above tip will not work.
```

Comment:
```suggestion
If you are bottlenecked on expensive CPU preprocessing and the preprocessed Dataset is too large to fit in object store memory, then the above tip will not work if your disk I/O is also too slow.
```

Reply:
I think we should just discourage this entirely? I think for pretty much all use cases, disk speed will not be fast enough for disk spilling to be worth it compared to adding more CPU nodes, is that not the case?

Reply:
I think that depends on cluster configuration. NVMe drives for example can support GBs/s.